Add sha/asm/keccak1600-mmx.pl, x86 MMX module.

dot-asm opened the pull request.

dot-asm commented:
Two commits are intentional. Original commit is kind of a reference...

richsalz commented on crypto/sha/asm/keccak1600-mmx.pl:
```diff
@@ -0,0 +1,442 @@
+# pointers to T[][] and A[][] and the end of round. Since number of
+# rounds is even last round writes to A[][] and everything works out.
+# It's argued that MMX is the only code path meaningful to implement
+# for x86. This is because non-MMX-capable processors is extincted
```
'extinct'

dot-asm replied:
Right. I think that "But older compiler is worse" came out wrong. I think I rather meant "Older compiler worse the result."

richsalz commented:
This is fine with me, but maybe @bernd-edlinger can offer some real review?  If not, take my +1

bernd-edlinger commented:
I'll have a look in the evening...

dot-asm commented:
Red cross from travis is unrelated. [I could have added [skip ci]]...

kaduk commented on crypto/sha/asm/keccak1600-mmx.pl:
```diff
@@ -23,17 +23,17 @@
 # rounds is even last round writes to A[][] and everything works out.
 # It's argued that MMX is the only code path meaningful to implement
-# for x86. This is because non-MMX-capable processors is extincted
+# for x86. This is because non-MMX-capable processors is extinct
```
"are an extinct", I think

kaduk commented on crypto/sha/asm/keccak1600-mmx.pl:
```diff
@@ -23,17 +23,17 @@
-# invest into fancy 32-bit code. And the decision doesn't seem to be
-# inadequate if one compares below results to "64-bit platforms in
+# processed byte on Pentium. Which is fair result. But older compiler,
+# worse the result. On the other hand one can wonder why not 128-bit
```
"But older compiler, worse the result" is a strange sentence fragment. Maybe,
"But older compilers produce worse results"?

dot-asm replied:
Well, compiler doesn't produce the result. They produce code which produces the result, but not the result. So shall we say "But older compilers produce worse code"? BTW, I wonder if "But older compiler *is*, worse the result" sound better?

dot-asm commented:
I've added another code update, improving Silvermont performance even further, to 33%. Hence this needs re-confirm. Sorry about that. It was "triggered" by the anomalously poor Silvermont performance, which turned to be frequency reporting problem. But there was even real improvement, which was two-fold. I replaced MSROM xchg and elongated some dependency chains.

bernd-edlinger commented on crypto/sha/asm/keccak1600-mmx.pl:
```diff
@@ -256,13 +253,16 @@ ()
+	 &xor	("edi","esi");		# &xchg	("esi","edi");
 	&movq	(&QWP(@D[1],"esp"),@C[1]);
+	 &xor	("esi","edi");
+	 &xor	("edi","esi");
```
You mean, these 3 insns are faster than xchg esi, edi ?

dot-asm replied:
As it turns out xchg is so-called MSROM instruction on Silvermont. And MSROM is kind of worst thing than can happen, because it stalls whole pipeline for few cycles and depending on which decoder port it ended up on it can incur additional penalty in top of "normal" overhead. In other words if MSROM instructions are inevitable, it takes careful scheduling, but best is if you can avoid them altogether. Of course question is how does alternative affect performance on *other* processors. And even more specifically by how much. It has to be by little. Like in this case. Losses on other processors are portion of percent (actually *none* on high-end processors, because not all resources are fully utilized), while on Silvermont it's several percent gain. So to answer the question, both yes and no. It's a trade-off solution, but it's weighed and conscious one.

bernd-edlinger replied:
Cool!
Hope gcc does not generate these xchg insns when they are so bad on this cpu type.
(Just in case, I happen to be also a gcc maintainer)

richsalz commented:
“The older the compiler, the worse the generated code is.”

richsalz commented:
+1

dot-asm commented:
Merged. Thanks.
